### PR TITLE
Statically link hiredis C library

### PR DESCRIPTION
**Pull request by zzak**
I want to statically link the hiredis library, but this isn't working for me.. please check it

### CODE REVIEW

File: mrbgem.rake [hunk 1]
```diff
@@ -4,5 +4,43 @@ MRuby::Gem::Specification.new('mruby-redis') do |spec|
   spec.version = '0.0.1'
   spec.linker.libraries << "hiredis"
```
Comment (matsumoto-r) [inline mrbgem.rake]: line 5: a compiler found the system libraries of hiredis before linking `"#{hiredis_dir}/libhiredis.a"`, I guess. Could you try it after removing `line 5`?

**Comment by zzak** [comment]
@matsumoto-r You can see it reproduced on latest travis build, what do you think?

**Comment by matsumoto-r** [comment]
Where is `include/hiredis/hiredis.h` when compiling hiredis? You should point the `include/` using `-I` option in `mrbgem.rake`.

**Comment by matsumoto-r** [comment]
@zzak `include` was created under current dir after `make PREFIX=`pwd` install`. 

```
$ pwd
/Users/matsumoto_r/DEV/hiredis
$ make PREFIX=`pwd` install
mkdir -p /Users/matsumoto_r/DEV/hiredis/include/hiredis /Users/matsumoto_r/DEV/hiredis/lib
cp -a hiredis.h async.h adapters /Users/matsumoto_r/DEV/hiredis/include/hiredis
cp -a libhiredis.dylib /Users/matsumoto_r/DEV/hiredis/lib/libhiredis.0.10.dylib
cd /Users/matsumoto_r/DEV/hiredis/lib && ln -sf libhiredis.0.10.dylib libhiredis.0.dylib
cd /Users/matsumoto_r/DEV/hiredis/lib && ln -sf libhiredis.0.dylib libhiredis.dylib
cp -a libhiredis.a /Users/matsumoto_r/DEV/hiredis/lib
$ ls -lR include/
total 0
drwxr-xr-x@ 5 matsumoto_r  staff  170  3  9 09:19 hiredis

include//hiredis:
total 40
drwxr-xr-x@ 5 matsumoto_r  staff   170  1  6  2013 adapters
-rw-r--r--@ 1 matsumoto_r  staff  4928  1  6  2013 async.h
-rw-r--r--@ 1 matsumoto_r  staff  8919  1  6  2013 hiredis.h

include//hiredis/adapters:
total 24
-rw-r--r--@ 1 matsumoto_r  staff  2605  1  6  2013 ae.h
-rw-r--r--@ 1 matsumoto_r  staff  2973  1  6  2013 libev.h
-rw-r--r--@ 1 matsumoto_r  staff  2366  1  6  2013 libevent.h
```

**Comment by zzak** [comment]
@matsumoto-r Everything is in `hiredis` directory of build dir:

```
~/../mruby-redis => ls -a mruby/build/host/mrbgems/mruby-redis/hiredis/
.                Makefile         dict.h           hiredis.pc       read.h
..               README.md        examples         libhiredis.a     read.o
.git             adapters         fmacros.h        libhiredis.dylib sds.c
.gitignore       async.c          hiredis-test     net.c            sds.h
.travis.yml      async.h          hiredis.c        net.h            sds.o
CHANGELOG.md     async.o          hiredis.h        net.o            test.c
COPYING          dict.c           hiredis.o        read.c           test.o
```

**Comment by matsumoto-r** [comment]
Did you run ``make PREFIX=`pwd` install`?

**Comment by zzak** [comment]
@matsumoto-r I get file not found:

```
gcc -g -std=gnu99 -O3 -Wall -Werror-implicit-function-declaration -Wdeclaration-after-statement -O3 -fPIC  -Wall -W -Wstrict-prototypes -Wwrite-strings -g -ggdb  -o hiredis-test   test.o libhiredis.a
Generating hiredis.pc for pkgconfig...
prefix=/Users/zzak/ales/mruby-redis/mruby/build/host/mrbgems/mruby-redis/hiredis
libdir=/Users/zzak/ales/mruby-redis/mruby/build/host/mrbgems/mruby-redis/hiredis
/bin/sh: /include/hiredis: No such file or directory
make: *** [hiredis.pc] Error 127
rake aborted!
make failed
```

I only changed `mrbgem.rake` to set prefix:

```ruby
  if ! File.exists? "#{hiredis_dir}/libhiredis.a"
    Dir.chdir hiredis_dir do
      e = {
        'CC' => "#{spec.build.cc.command} #{spec.build.cc.flags.join(' ')}",
        'CXX' => "#{spec.build.cxx.command} #{spec.build.cxx.flags.join(' ')}",
        'LD' => "#{spec.build.linker.command} #{spec.build.linker.flags.join(' ')}",
        'AR' => spec.build.archiver.command,
        'PREFIX' => `pwd`
      }

      run_command e, "make"
    end
  end
```

**Comment by zzak** [comment]
@matsumoto-r I fixed hiredis compile, but get the following errors when compiling `mrb_redis.c`:

```
/Users/zscott/code/mruby-redis/src/mrb_redis.c:509:78: error: too many arguments provided to function-like macro invocation
    mrb_value score_str = mrb_float_to_str(mrb, mrb_float_value(mrb, score), "%f");
                                                                             ^
/Users/zscott/code/mruby-redis/mruby/include/mruby/value.h:41:10: note: macro 'mrb_float_to_str' defined here
# define mrb_float_to_str(buf, i) sprintf(buf, "%.16e", i)
         ^
/Users/zscott/code/mruby-redis/src/mrb_redis.c:509:27: error: use of undeclared identifier 'mrb_float_to_str'
    mrb_value score_str = mrb_float_to_str(mrb, mrb_float_value(mrb, score), "%f");
                          ^
2 errors generated.
/Users/zscott/code/mruby-redis/src/mrb_redis.c:509:78: error: too many arguments provided to function-like macro invocation
    mrb_value score_str = mrb_float_to_str(mrb, mrb_float_value(mrb, score), "%f");
                                                                             ^
/Users/zscott/code/mruby-redis/mruby/include/mruby/value.h:41:10: note: macro 'mrb_float_to_str' defined here
# define mrb_float_to_str(buf, i) sprintf(buf, "%.16e", i)
         ^
/Users/zscott/code/mruby-redis/src/mrb_redis.c:509:27: error: use of undeclared identifier 'mrb_float_to_str'
    mrb_value score_str = mrb_float_to_str(mrb, mrb_float_value(mrb, score), "%f");
```

**Comment by zzak** [comment]
Ahh, mruby master fixes compile errors.

Now I need to resolve mrbtest link error:

```
>>> Test host <<<
dyld: Library not loaded: libhiredis.0.13.dylib
  Referenced from: /Users/zscott/code/mruby-redis/mruby/build/host/test/mrbtest
  Reason: image not found
```

File: src/mrb_redis.c [hunk 2]
```diff
@@ -29,7 +29,7 @@
 #include <stdlib.h>
 #include <string.h>
 #include <time.h>
-#include <hiredis/hiredis.h>
+#include "hiredis.h"
```
Comment (matsumoto-r) [inline src/mrb_redis.c]: When installing hiredis, hiredis create header files into `"#{hiredis_dir}/include"`. So, correct path is `hiredis/hiredis.h`.

File: mrbgem.rake [hunk 3]
```diff
@@ -2,7 +2,47 @@ MRuby::Gem::Specification.new('mruby-redis') do |spec|
...
+      }
+
+      run_command e, "make"
+      run_command e, "make install"
+    end
+  end
+
+  spec.cc.include_paths << hiredis_dir
+  spec.linker.library_paths << hiredis_dir
+
```
Comment (matsumoto-r) [inline mrbgem.rake]: If we use libhiredis.a for static link, `spec.cc` and `spec.linker` flag is the following configuration:

```
  spec.cc.include_paths << "#{hiredis_dir}/include"                              
  #spec.linker.library_paths << hiredis_dir                                      
  spec.linker.flags_before_libraries << "#{hiredis_dir}/lib/libhiredis.a"        
```

**Comment by matsumoto-r** [comment]
Hi, I reviewed your code. Could you check and modify these code and push to the branch?

**Comment by zzak** [comment]
@matsumoto-r I pushed your changes but I get the following error:

```
/usr/bin/ld: cannot find -lhiredis
```

File: mrbgem.rake [hunk 4]
```diff
@@ -2,7 +2,48 @@ MRuby::Gem::Specification.new('mruby-redis') do |spec|
   spec.license = 'MIT'
   spec.authors = 'MATSUMOTO Ryosuke'
   spec.version = '0.0.1'
-  spec.linker.libraries << "hiredis"
   # for expire test
+  require 'open3'
+
+  hiredis_dir = "#{build_dir}/hiredis"
+  spec.linker.libraries << "hiredis"
```
Comment (matsumoto-r) [inline mrbgem.rake]: When linking *.a file, remove `spec.linker.libraries << "hiredis"`.

File: mrbgem.rake [hunk 5]
```diff
@@ -2,7 +2,48 @@ MRuby::Gem::Specification.new('mruby-redis') do |spec|
...
+        'PREFIX' => hiredis_dir
+      }
+
+      run_command e, "make"
+      run_command e, "make install"
+    end
+  end
+
+  spec.cc.include_paths << "#{hiredis_dir}/include"
+  spec.linker.library_paths << "#{hiredis_dir}/lib"
```
Comment (matsumoto-r) [inline mrbgem.rake]: When not using `spec.linker.libraries << "hiredis"`, don't need `spec.linker.library_paths << "#{hiredis_dir}/lib"`.

**Comment by matsumoto-r** [comment]
@zzak I reviewed again. Please check it.

**Comment by matsumoto-r** [comment]
Hi, @zzak , this pull-request LGTM. Can I merge this?

**Comment by zzak** [comment]
@matsumoto-r Sounds good! Thank you!!!

**Pull request merged**